### PR TITLE
Draft: Intersect optimization for large unions

### PR DESCRIPTION
**Pull request by neclimdul**

### CODE REVIEW

File: src/Type/TypeCombinator.php [hunk 1]
```diff
@@ -638,11 +638,13 @@ public static function intersect(Type ...$types): Type
 			$topLevelUnionSubTypes = [];
 			$innerTypes = $type->getTypes();
 			usort($innerTypes, $sortTypes);
+			$slice1 = array_slice($types, 0, $i);
```
Comment (drupol) [inline src/Type/TypeCombinator.php:641]: Hi,

Just out of curiosity, why creating intermediary variables here?

Comment (staabm) [reply src/Type/TypeCombinator.php:641]: otherweise these arrays get `sliced` within the loop over and over again

Comment (drupol) [reply src/Type/TypeCombinator.php:641]: Oh you're right, for some reason I haven't seen the `foreach` loop ! Well done.

Comment (staabm) [reply src/Type/TypeCombinator.php:641]: @neclimdul I feel this `array_slice` change is a no-brainer and should be submitted as a seperate PR.
the other change breaks the tests atm, so needs further inspection

Comment (ondrejmirtes) [reply src/Type/TypeCombinator.php:641]: I agree, please send a microoptimization PR, then we'll discuss the rest.

**Comment by ondrejmirtes** [comment]
Ah whatever, I really like the optimization, the result can be slightly worse in this case :)

**Comment by ondrejmirtes** [comment]
I pushed test adjustments, will merge it if it comes back green.

**Comment by ondrejmirtes** [comment]
Thank you!

File: tests/PHPStan/Analyser/TypeSpecifierTest.php [hunk 2]
```diff
@@ -422,7 +422,7 @@ public function dataCondition(): array
 					new Variable('foo'),
 					new Variable('bar'),
 				),
-				['$foo' => 'Bar', '$bar' => 'Bar'],
+				['$foo' => 'Bar', '$bar' => 'mixed'], // could be '$bar' => 'Bar'
```
Comment (rvanvelzen) [inline tests/PHPStan/Analyser/TypeSpecifierTest.php:425]: This is really annoying to not have anymore :/

I might spend some time on this in the future to bring back.

Comment (ondrejmirtes) [reply tests/PHPStan/Analyser/TypeSpecifierTest.php:425]: I'm sorry, it looked pretty esoteric to me as a way to narrow down types :)

Comment (ondrejmirtes) [reply tests/PHPStan/Analyser/TypeSpecifierTest.php:425]: Maybe we can still do it only for objects, or something like that...

Comment (rvanvelzen) [reply tests/PHPStan/Analyser/TypeSpecifierTest.php:425]: If it annoys me enough I will find a way to bring it back without a huge performance loss :)

**Comment by herndlm** [comment]
I'm kind of travelling right now and could not check - do we know if this improves other big union operations too? E.g. some constant array perf problems also lead to many unions. Or is this a too different case?

**Comment by neclimdul** [comment]
@herndlm likely very specific.

**Comment by mvorisek** [comment]
@neclimdul this PR has introduced https://github.com/phpstan/phpstan/issues/7550 issue, will you fix it?

**Pull request merged**